galsim/chromatic.py

```diff
@@ -68,65 +112,50 @@ class ChromaticObject(object):
-    #    in order to draw (slow).
-    # 5) Separable objects must initialize an `SED` attribute, which is a callable object (often a
-    #    `galsim.SED` instance) that returns the _relative_ flux of the profile at a given
-    #    wavelength. (The _absolute_ flux is controlled by both the `SED` and the `.flux` attribute
-    #    of the underlying chromaticized GSObject(s).  See `galsim.Chromatic` docstring for details
-    #    concerning normalization.)
-    # 6) Initialize a `wave_list` attribute, which specifies wavelengths at which the profile (or
-    #    the SED in the case of separable profiles) will be evaluated when drawing a
-    #    ChromaticObject.  The type of `wave_list` should be a numpy array, and may be empty, in
-    #    which case either the Bandpass object being drawn against, or the integrator being used
-    #    will determine at which wavelengths to evaluate.
-
-    # Additionally, instances of `ChromaticObject` and subclasses will usually have either an `obj`
-    # attribute representing a manipulated `GSObject` or `ChromaticObject`, or an `objlist`
-    # attribute in the case of compound classes like `ChromaticSum` and `ChromaticConvolution`.
-
+    # ChromaticObjects should adhere to the following invariants:
+    # - Objects should define the attributes:
+    #   * .SED, .separable, .wave_list, .interpolated
+    # - obj.evaluateAtWavelength(lam).drawImage().array.sum() == obj.SED(lam)
+    #   == obj.evaluateAtWavelength(lam).getFlux()
+    # - if SED.spectral:
+    #       obj.SED.calculateFlux(bandpass) == obj.calculateFlux(bandpass)
+    #       == obj.drawImage(bandpass).array.sum()
+    # - .separable is a boolean indicating whether or not the profile can be factored into a
+    #   spatial part and a spectral part.
+    # - .wave_list is a numpy array indicating wavelengths of particular interest, for instance, the
+    #   wavelengths at which the SED is explicitly defined via a LookupTable.  These are the
+    #   wavelengths that will be used (in addition to those in bandpass.wave_list) when drawing an
+    #   image of the chromatic profile.
+    # - .interpolated is a boolean indicating whether any part of the object hierarchy includes an
+    #   InterpolatedChromaticObject.
 
     def __init__(self, obj):
+        from .deprecated import depr
+        depr('__init__', 1.5, '',
+             "Please apply wavelength-dependent transformations directly to GSObjects "
+             "instead of passing through ChromaticObject().")
+        self.separable = obj.separable
+        self.interpolated = obj.interpolated
+        self.wave_list = obj.wave_list
         self.obj = obj
         if isinstance(obj, galsim.GSObject):
-            self.separable = True
-            self.SED = galsim.SED('1', 'nm', 'flambda')  # uniform flambda = 1
-            self.wave_list = np.array([], dtype=float)
+            self.SED = galsim.SED(obj.flux, 'nm', '1')
```

Comment:
The thing that left me most confused is that
```
galsim.Convolve(gal*sed, psf).drawImage(...)
``
and
```
galsim.Convolve(galsim.ChromaticObject(gal) * sed, psf).drawImage(...)
```
end up producing essentially the same image, but have different SEDs along the way, differing by a factor of gal.flux.  I couldn't figure out why they both ended up with the same drawn flux.  It probably doesn't matter, since the code is correct, but it left me very confused.